kafka-connect-aws-s3/src/main/scala/io/lenses/streamreactor/connect/aws/s3/source/S3BucketReaderManager.scala

```diff
@@ -0,0 +1,156 @@
+/*
+ * Copyright 2020 Lenses.io
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.lenses.streamreactor.connect.aws.s3.source
+
+import com.typesafe.scalalogging.LazyLogging
+import io.lenses.streamreactor.connect.aws.s3.formats.S3FormatStreamReader
+import io.lenses.streamreactor.connect.aws.s3.model._
+import io.lenses.streamreactor.connect.aws.s3.source.config.SourceBucketOptions
+import io.lenses.streamreactor.connect.aws.s3.storage.StorageInterface
+import org.apache.kafka.common.errors.OffsetOutOfRangeException
+
+case class State(
+                  initFile: Option[S3StoredFile],
+                  initLine: Option[Int],
+                  file: Option[S3StoredFile],
+                  currentReader: Option[_ <: S3FormatStreamReader[_ <: SourceData]],
+                  terminated: Boolean = false
+                )
+
+/**
+  * Given a sourceBucketOptions, manages readers for all of the files
+  */
+class S3BucketReaderManager(
+                             sourceBucketOptions: SourceBucketOptions,
+                             offsetReaderResultFn: (String, String) => Option[OffsetReaderResult]
+                           )
+                           (
+                             implicit storageInterface: StorageInterface,
+                             sourceLister: S3SourceLister
+                           ) extends LazyLogging {
+
+  private val prefix = sourceBucketOptions.fileNamingStrategy.prefix(sourceBucketOptions.sourceBucketAndPrefix)
+  private val resultReader = new ResultReader(prefix, sourceBucketOptions.targetTopic)
+
+  private var state: State = setupInitialState()
+
+  def close(): Unit = state.currentReader.foreach(_.close)
+
+  def poll(): Vector[PollResults] = {
+    logger.debug(s"Fresh poll call received")
+
+    state = state.copy(terminated = false)
+    val limit = sourceBucketOptions.limit
+
+    var allResults: Vector[PollResults] = Vector()
+
+    var n = 0
+    var allLimit: Int = limit
+    do {
+      chooseReader() match {
+        case Some(currentReader) =>
+          val newResults = resultReader.retrieveResults(currentReader, allLimit)
+          state = if (newResults.isEmpty || (newResults.nonEmpty && newResults.get.resultList.size < allLimit)) {
```

Comment:
Best is to pattern match on the newResults than doing the forced .get

Reply:
Isn't this potentially taking a long time? 

If I have 1 000 000 files each with 1 000 000 records it could take hours to process them. The while loop terminates when all have been exhausted no?

Reply:
I'd be keen to discuss alternatives.